spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseReader.java

```diff
@@ -60,35 +63,57 @@
  *
  * @param <T> is the Java class returned by this reader whose objects contain one or more rows.
  */
-abstract class BaseDataReader<T> implements Closeable {
-  private static final Logger LOG = LoggerFactory.getLogger(BaseDataReader.class);
+abstract class BaseReader<T, TaskT extends ScanTask> implements Closeable {
+  private static final Logger LOG = LoggerFactory.getLogger(BaseReader.class);
 
   private final Table table;
-  private final Iterator<FileScanTask> tasks;
+  private final Schema expectedSchema;
+  private final boolean caseSensitive;
+  private final String nameMapping;
+  private final Iterator<TaskT> tasks;
   private final Map<String, InputFile> inputFiles;
 
   private CloseableIterator<T> currentIterator;
   private T current = null;
-  private FileScanTask currentTask = null;
+  private TaskT currentTask = null;
 
-  BaseDataReader(Table table, CombinedScanTask task) {
+  BaseReader(Table table, ScanTaskGroup<TaskT> task, Schema expectedSchema, boolean caseSensitive) {
```

Comment:
I think we should be able to get rid of changes in `ScanTask` and offer an abstract method like this:

```
protected abstract Stream<ContentFile<?>> referencedFiles(TaskT task);
```

We shouldn't be calling abstract methods in the constructor so you can make `inputFiles` lazy.

```
private final ScanTaskGroup<TaskT> taskGroup;
private final Iterator<TaskT> tasks;

...

private Map<String, InputFile> lazyInputFiles;

...

protected InputFile getInputFile(String location) {
  return inputFiles().get(location);
}

private Map<String, InputFile> inputFiles() {
  if (lazyInputFiles == null) {
    Stream<EncryptedInputFile> encryptedFiles = taskGroup.tasks().stream()
        .flatMap(this::referencedFiles)
        .map(file -> EncryptedFiles.encryptedInput(table.io().newInputFile(file.path().toString()), file.keyMetadata()));

    // decrypt with the batch call to avoid multiple RPCs to a key server, if possible
    Iterable<InputFile> decryptedFiles = table.encryption().decrypt(encryptedFiles::iterator);

    Map<String, InputFile> files = Maps.newHashMapWithExpectedSize(taskGroup.tasks().size());
    decryptedFiles.forEach(decrypted -> files.putIfAbsent(decrypted.location(), decrypted));
    this.lazyInputFiles = ImmutableMap.copyOf(files);
  }

  return lazyInputFiles;
}
```